contracts/test/subcall.ts

```diff
@@ -1,28 +1,184 @@
+  owner: SignerWithAddress,
+) {
+  const balance = await contract.provider.getBalance(contract.address);
+  if (balance.lt(initialBalance)) {
+    const resp = await owner.sendTransaction({
+      to: contract.address,
+      value: initialBalance.sub(balance),
+      data: '0x',
+    });
+    await resp.wait();
+  }
+  expect(await contract.provider.getBalance(contract.address)).eq(
+    initialBalance,
+  );
+}
 
-import { Decoded, bech32 } from 'bech32';
+function decodeResult(receipt: ContractReceipt) {
+  const event = receipt.events![0].args! as unknown as {
+    status: number;
+    data: string;
+  };
+  return {
+    status: event.status,
+    data:
+      event.status == 0
+        ? cborg.decode(ethers.utils.arrayify(event.data))
+        : new TextDecoder().decode(ethers.utils.arrayify(event.data)),
+  };
+}
 
 describe('Subcall', () => {
   let contract: SubcallTests;
-  let alice: Decoded;
+  let owner: SignerWithAddress;
+  let ownerAddr: string;
+  let ownerNativeAddr: Uint8Array;
+  let kp: { publicKey: Uint8Array; secretKey: Uint8Array };
+
   before(async () => {
-    alice = bech32.decode('oasis1qrec770vrek0a9a5lcrv0zvt22504k68svq7kzve');
     const factory = await ethers.getContractFactory('SubcallTests');
-    contract = (await factory.deploy()) as SubcallTests;
+    contract = (await factory.deploy({
+      value: parseEther('1.0'),
+    })) as SubcallTests;
+
+    const signers = await ethers.getSigners();
+    owner = signers[0];
+    ownerAddr = await owner.getAddress();
+
+    // Convert Ethereum address to native bytes with version prefix (V1=0x00)
+    ownerNativeAddr = ethers.utils.arrayify(
+      ethers.utils.zeroPad(ownerAddr, 21),
+    );
+    expect(ownerNativeAddr.length).eq(21);
+
+    const rawKp = await contract.generateRandomAddress();
+    kp = {
+      publicKey: ethers.utils.arrayify(rawKp.publicKey),
+      secretKey: ethers.utils.arrayify(rawKp.secretKey),
+    };
   });
-  it('account.Transfer', async () => {
-    const addr = new Uint8Array(21);
-    const transfer = cborg.encode({
-      to: new Uint8Array(alice.words.slice(1)),
-      amount: 0n,
-    });
-    const tx = await contract.testSubcall('account.Transfer', transfer);
-    const receipt = await tx.wait();
-    const event = receipt.events![0].args!;
-    expect(event.status).eq(3);
-    expect(event.data).eq('0x636f7265');
+
+  it('Derive Staking Addresses', async () => {
+    const newKeypair = await contract.generateRandomAddress();
+
+    // Verify @oasisprotocol/client matches Solidity
+    const alice = oasis.signature.NaclSigner.fromSeed(
+      ethers.utils.arrayify(newKeypair.secretKey),
+      'this key is not important',
+    );
+    const computedPublicKey = ethers.utils.hexlify(
+      await oasis.staking.addressFromPublicKey(alice.public()),
+    );
+
+    expect(computedPublicKey).eq(ethers.utils.hexlify(newKeypair.publicKey));
+  });
+
+  it('accounts.Transfer', async () => {
+    // Ensure contract has an initial balance
```

Comment:
```suggestion
    // Ensure contract has an initial balance.
```

Please add full-stops at the end of regular comments. Also below.